Drafts authentication chapter

chrisbay opened the pull request:
Still needs concept checks

chrisbay committed:
WIP on auth stuff

chrisbay committed:
Merge branch 'master' into auth

chrisbay committed:
Completes conceptual sections of auth chapter

chrisbay committed:
Drafts login/registration form section

gildedgardenia committed:
Added authentication flow diagram

chrisbay committed:
Completes draft of authentication chapter

chrisbay committed:
Merge branch 'auth' of github.com:LaunchCodeEducation/java-web-development into auth

CarlyLanglois commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,97 @@
+   At some point, you may have cleared all of the cookies from your browser only to find that you were suddenly logged out of every website on the Internet. This is because those cookies stored the session IDs that allowed the various sites to know who you were. 
+
+A logout request to a web app simply destroys the user's session, erasing any memory of who they are. Even if they pass a session cookie in their subsequent requests, the corresponding session and session ID will no longer exist. 
+
```
For the second sentence, what about:

"Even if they pass a new session cookie in their subsequent requests, the older session and session ID no longer exist."

Using corresponding makes me unsure if you're referring to a correspondence with the subsequent request

CarlyLanglois commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,97 @@
+----------------------------------
+
+Before we proceed, we want to point out something important about authentication in Spring. Spring contains a sub-project, `Spring Security <https://spring.io/projects/spring-security>`_, that provides extensive support for authentication *and* authorization. In addition to supporting simple authentication, Spring Security also supports more sophisticated flows like `OAuth 2 <https://oauth.net/2/ ...
+
```
sophisticated flows -> sophisticated authorization flows/processes.
some students might not be used to referring to the auth process as a flow?

CarlyLanglois commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,97 @@
+In this book, we are explicitly not introducing Spring Security for two reasons:
+
+#. The project handles many aspects of the authentication process for you. This hides many of the steps that are important to understand as you learn about authentication.
+#. Setting up Spring Security is fairly complicated, and requires concept that are beyond the scope of this course.
```
requires concepts*

CarlyLanglois commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,97 @@
+#. The project handles many aspects of the authentication process for you. This hides many of the steps that are important to understand as you learn about authentication.
+#. Setting up Spring Security is fairly complicated, and requires concept that are beyond the scope of this course.
+
+That said, the authentication approach outlined in this chapter is sufficient for use in your personal projects. And never fear, when you begin working with a team on professional application, a senior developer will have already set up authentication.
```
on a* professional application

CarlyLanglois replied:
actually, that last sentence is kind of a black box :/ maybe we can say something like: 

"When you begin working with a team on professional applications, a senior developer will likely be on hand to help with authentication setup."

CarlyLanglois commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,97 @@
+
+The logical flow of simple authentication looks like this:
+
+.. figure:: figures/authenticationflowdiagram.png
+   :alt: A diagram showing the the flow of a username and password through the authentication process.
```
caption missing

CarlyLanglois committed:
ccs page 1

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+
+.. _hash_properties:
+
+#. **Deterministic**: If we encode a message with the function a different points in time then we always get the same result.
```
at different points*?

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+
+#. **Deterministic**: If we encode a message with the function a different points in time then we always get the same result.
+#. **Hard to reverse**: It is infeasible to calculate the input value that yields a given hash.
+#. **Hash values are ALMOST unique**: If ``a`` and ``b`` are two different messages, then is is *extremely* unlikely that they have the same hash value. By "extremely unlikely" we mean that this is something like a one-in-a-trillion likelihood (there are only about 8 billion people on earth). 
```
I'd add a transition into the parentheses - for context, ... eg

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+Password Verification With Hashes
+---------------------------------
+
+Our application will select a particular hash function. Let's call it ``h``. Then, for a message ``x``, the hash value will be the result of call ``h`` with the argument ``x``: ``h(x)``.  Rather than store passwords in a database, we will store their hash values.
```
I looked at this as an argument of x:h(x). what about:

"Then, for a message ``x``, the hash value will be the result of call ``h`` with the argument ``x``. Invoked, this looks like: ``h(x)``."

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+      // the hashes are different, so the passwords are definitely different
+   }
+
+The conditional compares the values of the hash stored in the database with the hash generated from the *submitted* password. By :ref:`property 1 <hash_properties>` we know that if the hash values are different, then there is no way the passwords are the same. By :ref:`property 3 <hash_properties>` we can safely assume that the passwords are the same. 
```
By :ref:`property 1 <hash_properties>`, - (add a comma)

CarlyLanglois replied:
same with the last sentence - comma after "property 3"

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+
+.. index:: ! collision
+
+Using hash functions to process passwords is not a cure-all. One vulnerability is the possibility for **collisions**. A collision occurs when two different messages have the same hash value. By :ref:`property 3 <hash_properties>`, this is supposed to be rare. However, if a collision is found for a given hash function then it may be possible to create an algorithm to *generate* collisions. In othe ...
```
for a given hash function, - (add a comma)

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+   The widespread use of brute force attacks is why it is always a bad idea to use a password that:
+
+   - is ranked as one of the `most commonly used passwords <https://en.wikipedia.org/wiki/List_of_the_most_common_passwords>`_,
+   - utilizes publicly accessible information about you---such as birth date or address---or
```
Jim usually has me remove the conjunctions when using a list

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+   - utilizes publicly accessible information about you---such as birth date or address---or
+   - uses common words from the dictionary.
+
+   When trying to crack a password hash using brute force, these are first items a hacker will attempt to use.
```
the first items*

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+
+Fortunately, it is possible for us to implement simple authentication *without* storing passwords, by using a technique called **password hashing**. Password hashing is a particular type of encryption that we will explore throughout the rest of this section.
+
+One-Way Vs. Two-Way Encryption
```
I think making these two separate headings could be useful to emphasize that hashing is one-way encrypt. and that this is what we'll practice in this class.

I think you and Sally are talking about this offline but I think if we have a two way diagram, we should also have a one way diagram or point out that one way is simply the top half of the given image. 
Since it is introduced juxtaposed to two way encrypt., I think some more upfront explanation of how hashing works is needed. It becomes clearer by the time you reach the example code block how hashing does its job, but before then, we're missing an explanation of how a message is de-crypted in one-way encryption - or rather, an explanation of how one-way encryption functions in the absence of decryption. 
Another note on this section, I think it would be helpful to add some language - maybe before going over the properties of hashing - around introducing hashing algorithms as code written by others. Otherwise, some students may read most of the page thinking they will be responsible for writing such a tool

CarlyLanglois replied:
also maybe just calling out the idea of a one-way function might help highlight the hash mechanism

chrisbaylewis replied:
Updated. Let me know what you think!

CarlyLanglois replied:
yeah very clear now!

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,123 @@
+
+The once-popular MD5 and SHA1 hash algorithms `quickly become obsolete <https://arstechnica.com/information-technology/2017/02/at-deaths-door-for-years-widely-used-sha1-function-is-now-dead/>`_ (for cryptographic purposes, at least) once collisions were found. 
+
+Another vulnerability of hashing is the growth in computing power. If a hacker breaks into a database, they will obtain the hashes of all of its users' passwords. Since only a small handful of hash functions are commonly used, they might simply try millions of strings with each of the more popular hash functions and wait until they find a match. 
```
I might rephrase the first sentence here to something like,

"Hashing becomes more vulnerable as global computing power increases."

CarlyLanglois committed:
add ccs to hashing page

CarlyLanglois committed:
fix styling

CarlyLanglois commented on src/chapters/auth/user-model.rst:
```diff
@@ -0,0 +1,122 @@
+
+   }
+
+Notice that the constructor takes a parameter named ``password`` and uses it to set the value of ``pwHash``. We mentioned :ref:`previously <hashing-passwords>` that we should never store passwords, so in a moment we will create a hash from the given password to store.
```
"so in a moment, we will update line 26 by creating a hash from the given password to store."

CarlyLanglois committed:
ccs for user model

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+Creating ``AuthenticationController``
+-------------------------------------
+
+In the ``controllers`` package, create a new class named ``AuthenticationController``. Since this controller will deal with ``User`` objects, it needs a ``UserRepository``.
```
a UserRepository instance?

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+Session-Handling Utilities
+--------------------------
+
+Before creating handler methods for rendering and processing our login and registration forms, we need some utility methods for working with sessions. Below ``userRepository``, let's add the following class members:
```
below the definition of ``userRepository``, ...

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+
+``LoginFormDTO`` should live in the ``models.dto`` package.
+
+.. admonition:: Tip
```
I think this tip is well placed and should help solidify the ideas introduced in the last chapter - i'd add a nod to this like, 
"remember, we can think of a dto ..." something to tie it in with the introduction of this tool

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+.. sourcecode:: java
+   :lineno-start: 6
+
+   public class RegisterFormDTO extends LoginFormDTO {
```
very minute so take it or leave it, but this could be registrationFormDTO - as it is now, some items use register and others use registration

chrisbaylewis replied:
Valid point. I'll probably leave it as-is though.

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+
+      if (errors.hasErrors()) {
+         model.addAttribute("title", "Register");
+         model.addAttribute("title", "Register");
```
duplicate line

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+- **Lines 60-64**: Return the user to the form if an validation errors occur.
+- **Line 66**: Retrieve the user with the given username from the database.
+- **Lines 68-72**: If a user with the given username already exists, register a custom error with the ``errors`` object and return the user to the form. See the note on using ``errors.rejectValue`` below.
+- **Lines 74-80**: Compare the two passwords submitted. If they do not match, register a custom error adn return the user to the form.
```
and*

CarlyLanglois committed:
ccs for login/reg forms page

CarlyLanglois commented on src/chapters/auth/request-filtering.rst:
```diff
@@ -0,0 +1,182 @@
+   :lineno-start: 27
+
+   @Override
+   public boolean preHandle(HttpServletRequest request,
```
I'm not familiar with IOException but the doc site we send them to just throws Exception. we may want to clarify that these don't need to match exactly? or do they?

chrisbaylewis replied:
Correct on the docs. However, the `sendRedirect`call we use near the bottom throws IOException, which is a checked exception, so we have either handle or re-throw.

CarlyLanglois replied:
I think since we emphasize that the signatures should match exactly and send them to the method they are overriding, we should clarify that its okay that the throw statement does not match

chrisbaylewis replied:
Good call.

CarlyLanglois committed:
cc questions for last page

CarlyLanglois commented on src/chapters/auth/request-filtering.rst:
```diff
@@ -0,0 +1,182 @@
+
+The first method---``authenticationFilter``, which is annotated with ``@Bean``---makes our filter available as a Spring-managed class. 
+
+The code in ``WebApplicationConfig`` is hard to understand, and contains some advanced concepts that you haven't fully learned. That's okay. We don't expect you to understand every detail of this class. However, you do need to understand its role in enabling authentication, which is to register our filter with Spring.
```
Is there a task we can ask them to do once they get to the end of this page to test their new filter functionality?

Also, this caveat is helpful here but could also be stated differently as a disclaimer earlier in this section - there is a lot of new material happening that is presented more quickly than we have been doing previously in the site

CarlyLanglois commented on src/chapters/auth/login-registration-forms.rst:
```diff
@@ -0,0 +1,378 @@
+         this.username = username;
+      }
+
+      public String getPassword() {
+         return password;
+      }
+
+      public void setPassword(String password) {
+         this.password = password;
+      }
+
```
I'm a little lost on what values these fields are for - are they the values from the form fields?

CarlyLanglois replied:
actually, i just re-read the tip below and answered my own question - but we may find it helpful to call this idea out more specifically for this instance because it is right after drilling home the idea to not store password values

chrisbaylewis replied:
I added a note about the `password` field. Let me know what you think.

CarlyLanglois replied:
yeah thats great - although you should add a comma after the first clause

CarlyLanglois reviewed (commented):
No typos that i spotted on the last two pages :) overall, there is a lot of new content in here! In the future, it could be nice to add some more theory/explanation around some of the new methodology introduced here - specifically the last couple pages. But I think for the course now, we could help manage expectations for students perhaps with more words upfront about the fact that we are about to introduce a number of new tools and packages in a faster manner than we have before.
this could also be a future task but a diagram to summarize this process could help contextualize the flow before we have the students dive into the dirty work - and summarize it one they have reached the end of the walkthrough.

chrisbay commented:
I addressed everything mentioned. Unresolved comments are items that I felt could use a re-read to verify that the new version works. I also moved my "this is hard stuff" caveat from the end of the chapter to the beginning of the user section.

chrisbay committed:
Improvements based on PR review

chrisbaylewis commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,147 @@
+
+   Select all that are true about salting:
+
+   #. Salting is an alternative to a hashing function.
```
Since salting is only described in an external resource, and isn't part of the learning objectives, I'm hesitant to include a question on it.

Also, a question reinforcing how hashing still allows for submitted passwords to be verified would be good.

chrisbaylewis commented on src/chapters/auth/request-filtering.rst:
```diff
@@ -0,0 +1,204 @@
+
+.. admonition:: Question
+
+   True/False: Paths included in a whitelist must be exact.
```
The way this is phrased is a little confusing. And we should specify that this is how *our* whitelist code works (since there's not standard behavior here). How about: "When our code checks a path against entries in the whitelist, it must match exactly in order for it be accessed without logging in."

chrisbay committed:
Adds links to code repo branches

chrisbay committed:
Fixes broken link

CarlyLanglois commented on src/chapters/auth/hashing-passwords.rst:
```diff
@@ -0,0 +1,150 @@
+
+**One-way encryption**, also known as **hashing**, encodes a message in a way that makes it *nearly* impossible to decode. An algorithm implementing a one-way encryption algorithm is a **cryptographic hash function** (we'll usually leave off "cryptographic"). For a specific message, the output of a hash function is a **hash**. 
+
+The hashing process is essentially just the top half of the diagram above. While it might not be clear yet, hashing allows us to securely store passwords and still be able to validate a user's submitted password. Precisely _how_ this is done will be explored in a moment. 
```
*how*?

CarlyLanglois replied:
as in, did you mean for this to italicized?

CarlyLanglois committed:
remove salting cc, add a ahshing desc question

CarlyLanglois committed:
fix whitelist question

chrisbay committed:
Minor improvements based on review

chrisbay committed:
Merge branch 'auth' of github.com:LaunchCodeEducation/java-web-development into auth

gildedgardenia committed:
First sessions and cookies diagram

gildedgardenia committed:
Merge branch 'auth' of https://github.com/LaunchCodeEducation/java-web-development into auth

gildedgardenia committed:
Added diagrams to text

gildedgardenia commented on src/chapters/auth/intro.rst:
```diff
@@ -0,0 +1,133 @@
+   single: session; ID
+
+#. When a user is authenticated---that is, they provide a valid username/password pair---a session containing their user ID is created on the server. The session has a unique **session ID**, such as ``81LfWG9``.
+#. In order to associate the user with their session, the server returns a **session cookie** to the browser within an HTTP response. The session contains the session ID, and looks something like this: ``session_id: 81LfWG9``. 
```
I think this step should be clarified. "..., the server creates a **session cookie** and sends the cookie to the browser...". Also, it should say "The session *cookie* contains the session ID..."

gildedgardenia reviewed (commented):
Just thought one thing was kind of confusing!

chrisbay committed:
Fixes wording on session cookies

chrisbay committed:
Merge branch 'auth' of github.com:LaunchCodeEducation/java-web-development into auth

gildedgardenia committed:
Added captions

CarlyLanglois committed:
fix rst typo

CarlyLanglois committed:
Merge branch 'auth' of github.com:LaunchCodeEducation/java-web-development into auth
